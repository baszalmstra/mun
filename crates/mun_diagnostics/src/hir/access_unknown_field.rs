use super::HirDiagnostic;
use crate::{Diagnostic, SourceAnnotation};
use mun_hir::HirDisplay;
use mun_syntax::{ast, AstNode, TextRange};

/// An error that is emitted when trying to access a field that doesn't exist.
///
/// ```mun
/// struct Foo {
///     b: i32
/// }
///
/// # fn main() {
/// let a = Foo { b: 3}
/// let b = a.c;    // no field `c`
/// #}
/// ```
pub struct AccessUnknownField<'db, 'diag> {
    db: &'db dyn mun_hir::HirDatabase,
    diag: &'diag mun_hir::diagnostics::AccessUnknownField,
    location: TextRange,
}

impl<'db, 'diag> Diagnostic for AccessUnknownField<'db, 'diag> {
    fn range(&self) -> TextRange {
        self.location
    }

    fn title(&self) -> String {
        format!(
            "no field `{}` on type `{}`",
            self.diag.name,
            self.diag.receiver_ty.display(self.db),
        )
    }

    fn primary_annotation(&self) -> Option<SourceAnnotation> {
        Some(SourceAnnotation {
            range: self.location,
            message: "unknown field".to_string(),
        })
    }
}

impl<'db, 'diag> AccessUnknownField<'db, 'diag> {
    /// Constructs a new instance of `AccessUnknownField`
    pub fn new(
        db: &'db dyn mun_hir::HirDatabase,
        diag: &'diag mun_hir::diagnostics::AccessUnknownField,
    ) -> Self {
        let parse = db.parse(diag.file);

        let location = ast::FieldExpr::cast(diag.expr.to_node(&parse.syntax_node()))
            .map(|f| f.field_range())
            .unwrap_or_else(|| diag.highlight_range());

        AccessUnknownField { db, diag, location }
    }
}
