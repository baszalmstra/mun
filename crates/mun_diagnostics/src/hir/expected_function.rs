use super::HirDiagnostic;
use crate::{Diagnostic, SourceAnnotation};
use mun_hir::HirDisplay;
use mun_syntax::TextRange;

/// An error that is emitted when a function is expected but something else is encountered:
///
/// ```mun
/// # fn main() {
///     let a = 3;
///     let b = a();    // expected function
/// # }
/// ```
pub struct ExpectedFunction<'db, 'diag> {
    db: &'db dyn mun_hir::HirDatabase,
    diag: &'diag mun_hir::diagnostics::ExpectedFunction,
}

impl<'db, 'diag> Diagnostic for ExpectedFunction<'db, 'diag> {
    fn range(&self) -> TextRange {
        self.diag.highlight_range()
    }

    fn title(&self) -> String {
        format!(
            "expected function, found `{}`",
            self.diag.found.display(self.db)
        )
    }

    fn primary_annotation(&self) -> Option<SourceAnnotation> {
        Some(SourceAnnotation {
            range: self.diag.highlight_range(),
            message: "not a function".to_owned(),
        })
    }
}

impl<'db, 'diag> ExpectedFunction<'db, 'diag> {
    /// Constructs a new instance of `ExpectedFunction`
    pub fn new(
        db: &'db dyn mun_hir::HirDatabase,
        diag: &'diag mun_hir::diagnostics::ExpectedFunction,
    ) -> Self {
        ExpectedFunction { db, diag }
    }
}
