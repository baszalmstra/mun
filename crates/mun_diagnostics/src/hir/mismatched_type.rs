use super::HirDiagnostic;
use crate::{Diagnostic, SourceAnnotation};
use mun_hir::HirDisplay;
use mun_syntax::TextRange;

/// An error that is emitted when a different type was found than expected.
///
/// ```mun
/// fn add(a: i32, b: i32) -> i32{
///     a+b
/// }
///
/// # fn main() {
///     add(true, false); // type mismatch, expected i32 found bool.
/// # }
/// ```
pub struct MismatchedType<'db, 'diag> {
    db: &'db dyn mun_hir::HirDatabase,
    diag: &'diag mun_hir::diagnostics::MismatchedType,
}

impl<'db, 'diag> Diagnostic for MismatchedType<'db, 'diag> {
    fn range(&self) -> TextRange {
        self.diag.highlight_range()
    }

    fn title(&self) -> String {
        format!(
            "expected `{}`, found `{}`",
            self.diag.expected.display(self.db),
            self.diag.found.display(self.db)
        )
    }

    fn primary_annotation(&self) -> Option<SourceAnnotation> {
        None
    }
}

impl<'db, 'diag> MismatchedType<'db, 'diag> {
    /// Constructs a new instance of `MismatchedType`
    pub fn new(
        db: &'db dyn mun_hir::HirDatabase,
        diag: &'diag mun_hir::diagnostics::MismatchedType,
    ) -> Self {
        MismatchedType { db, diag }
    }
}
