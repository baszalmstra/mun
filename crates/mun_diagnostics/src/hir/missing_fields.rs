use super::HirDiagnostic;
use crate::{Diagnostic, SourceAnnotation};
use mun_hir::HirDisplay;
use mun_syntax::{ast, AstNode, TextRange};

/// An error that is emitted when a field is missing from a struct initializer.
///
/// ```mun
/// struct Foo {
///     a: i32,
/// }
///
/// # fn main() {
///     let a = Foo {}; // missing field `a`
/// # }
/// ```
pub struct MissingFields<'db, 'diag> {
    db: &'db dyn mun_hir::HirDatabase,
    diag: &'diag mun_hir::diagnostics::MissingFields,
    location: TextRange,
    missing_fields: String,
}

impl<'db, 'diag> Diagnostic for MissingFields<'db, 'diag> {
    fn range(&self) -> TextRange {
        self.location
    }

    fn title(&self) -> String {
        format!(
            "missing fields {} in initializer of `{}`",
            self.missing_fields,
            self.diag.struct_ty.display(self.db)
        )
    }

    fn primary_annotation(&self) -> Option<SourceAnnotation> {
        Some(SourceAnnotation {
            range: self.location,
            message: format!("missing {}", self.missing_fields.clone()),
        })
    }
}

impl<'db, 'diag> MissingFields<'db, 'diag> {
    /// Constructs a new instance of `MissingFields`
    pub fn new(
        db: &'db dyn mun_hir::HirDatabase,
        diag: &'diag mun_hir::diagnostics::MissingFields,
    ) -> Self {
        let parse = db.parse(diag.file);
        let missing_fields = diag
            .field_names
            .iter()
            .map(|n| format!("`{}`", n))
            .collect::<Vec<String>>()
            .join(", ");

        let location = ast::RecordLit::cast(diag.fields.to_node(&parse.syntax_node()))
            .and_then(|f| f.type_ref())
            .map(|t| t.syntax().text_range())
            .unwrap_or_else(|| diag.highlight_range());

        MissingFields {
            db,
            diag,
            location,
            missing_fields,
        }
    }
}
