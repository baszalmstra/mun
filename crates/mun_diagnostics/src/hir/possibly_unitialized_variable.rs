use super::HirDiagnostic;
use crate::{Diagnostic, SourceAnnotation};
use mun_syntax::TextRange;

/// An error that is emitted when trying to access a field that is potentially not yet initialized.
///
/// ```mun
/// # fn main() {
/// let a;
/// let b = a;    // `a` is possible not yet initialized
/// #}
/// ```
pub struct PossiblyUninitializedVariable<'db, 'diag> {
    _db: &'db dyn mun_hir::HirDatabase,
    diag: &'diag mun_hir::diagnostics::PossiblyUninitializedVariable,
    value_name: String,
}

impl<'db, 'diag> Diagnostic for PossiblyUninitializedVariable<'db, 'diag> {
    fn range(&self) -> TextRange {
        self.diag.highlight_range()
    }

    fn title(&self) -> String {
        format!("use of possibly-uninitialized `{}`", self.value_name)
    }

    fn primary_annotation(&self) -> Option<SourceAnnotation> {
        None
    }
}

impl<'db, 'diag> PossiblyUninitializedVariable<'db, 'diag> {
    /// Constructs a new instance of `PossiblyUninitializedVariable`
    pub fn new(
        db: &'db dyn mun_hir::HirDatabase,
        diag: &'diag mun_hir::diagnostics::PossiblyUninitializedVariable,
    ) -> Self {
        let parse = db.parse(diag.file);

        // Get the text of the value as a string
        let value_name = diag.pat.to_node(&parse.syntax_node()).text().to_string();

        PossiblyUninitializedVariable {
            _db: db,
            diag,
            value_name,
        }
    }
}
