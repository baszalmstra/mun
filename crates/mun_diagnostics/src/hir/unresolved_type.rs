use super::HirDiagnostic;
use crate::{Diagnostic, SourceAnnotation};
use mun_syntax::{AstNode, TextRange};

/// An error that is emitted when trying to use a type that doesnt exist within the scope.
///
/// ```mun
/// # fn main() {
/// let a = DoesntExist {}; // Cannot find `DoesntExist` in this scope.
/// #}
/// ```
pub struct UnresolvedType<'db, 'diag> {
    _db: &'db dyn mun_hir::HirDatabase,
    diag: &'diag mun_hir::diagnostics::UnresolvedType,
    value_name: String,
}

impl<'db, 'diag> Diagnostic for UnresolvedType<'db, 'diag> {
    fn range(&self) -> TextRange {
        self.diag.highlight_range()
    }

    fn title(&self) -> String {
        format!("cannot find type `{}` in this scope", self.value_name)
    }

    fn primary_annotation(&self) -> Option<SourceAnnotation> {
        Some(SourceAnnotation {
            range: self.diag.highlight_range(),
            message: "not found in this scope".to_owned(),
        })
    }
}

impl<'db, 'diag> UnresolvedType<'db, 'diag> {
    /// Constructs a new instance of `UnresolvedType`
    pub fn new(
        db: &'db dyn mun_hir::HirDatabase,
        diag: &'diag mun_hir::diagnostics::UnresolvedType,
    ) -> Self {
        let parse = db.parse(diag.file);

        // Get the text of the value as a string
        let value_name = diag
            .type_ref
            .to_node(&parse.syntax_node())
            .syntax()
            .text()
            .to_string();

        UnresolvedType {
            _db: db,
            diag,
            value_name,
        }
    }
}
