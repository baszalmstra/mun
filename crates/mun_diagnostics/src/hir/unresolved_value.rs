use super::HirDiagnostic;
use crate::{Diagnostic, SourceAnnotation};
use mun_syntax::{AstNode, TextRange};

/// An error that is emitted when trying to use a value that doesnt exist within the scope.
///
/// ```mun
/// # fn main() {
/// let a = b; // Cannot find `b` in this scope.
/// #}
/// ```
pub struct UnresolvedValue<'db, 'diag> {
    _db: &'db dyn mun_hir::HirDatabase,
    diag: &'diag mun_hir::diagnostics::UnresolvedValue,
    value_name: String,
}

impl<'db, 'diag> Diagnostic for UnresolvedValue<'db, 'diag> {
    fn range(&self) -> TextRange {
        self.diag.highlight_range()
    }

    fn title(&self) -> String {
        format!("cannot find value `{}` in this scope", self.value_name)
    }

    fn primary_annotation(&self) -> Option<SourceAnnotation> {
        Some(SourceAnnotation {
            range: self.diag.highlight_range(),
            message: "not found in this scope".to_owned(),
        })
    }
}

impl<'db, 'diag> UnresolvedValue<'db, 'diag> {
    /// Constructs a new instance of `UnresolvedValue`
    pub fn new(
        db: &'db dyn mun_hir::HirDatabase,
        diag: &'diag mun_hir::diagnostics::UnresolvedValue,
    ) -> Self {
        let parse = db.parse(diag.file);

        // Get the text of the value as a string
        let value_name = diag.expr.to_node(&parse.tree().syntax()).text().to_string();

        UnresolvedValue {
            _db: db,
            diag,
            value_name,
        }
    }
}
