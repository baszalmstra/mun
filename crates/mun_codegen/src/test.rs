use crate::{mock::MockDatabase, IrDatabase};
use hir::{diagnostics::DiagnosticSink, line_index::LineIndex, Module, SourceDatabase};
use inkwell::OptimizationLevel;
use std::cell::RefCell;
use std::sync::Arc;

#[test]
fn function() {
    test_snapshot(
        r#"
    fn main() {
    }
    "#,
    );
}

#[test]
fn return_type() {
    test_snapshot(
        r#"
    fn main():int {
      0
    }
    "#,
    );
}

#[test]
fn function_arguments() {
    test_snapshot(
        r#"
    fn main(a:int):int {
      a
    }
    "#,
    );
}

#[test]
fn binary_expressions() {
    test_snapshot(
        r#"
    fn add(a:int, b:int):int {
      a+b
    }

    fn subtract(a:int, b:int):int {
      a-b
    }

    fn multiply(a:int, b:int):int {
      a*b
    }
    "#,
    );
}

#[test]
fn let_statement() {
    test_snapshot(
        r#"
    fn main(a:int):int {
      let b = a+1
      b
    }
    "#,
    );
}

#[test]
fn invalid_binary_ops() {
    test_snapshot(
        r#"
    fn main() {
      let a = 3+3.0;
      let b = 3.0+3;
    }
    "#,
    );
}

#[test]
fn update_operators() {
    test_snapshot(
        r#"
    fn add(a:int, b:int):int {
      let result = a
      result += b
      result
    }

    fn subtract(a:int, b:int):int {
      let result = a
      result -= b
      result
    }

    fn multiply(a:int, b:int):int {
      let result = a
      result *= b
      result
    }

    fn divide(a:int, b:int):int {
      let result = a
      result /= b
      result
    }
    "#,
    );
}

#[test]
fn update_parameter() {
    test_snapshot(
        r#"
    fn add_three(a:int):int {
      a += 3;
      a
    }
    "#,
    );
}

#[test]
fn function_calls() {
    test_snapshot(
        r#"
    fn add_impl(a:int, b:int):int {
        a+b
    }

    fn add(a:int, b:int):int {
      add_impl(a,b)
    }

    fn test():int {
      add(4,5)
      add_impl(4,5)
      add(4,5)
    }
    "#,
    );
}

#[test]
fn equality_operands() {
    test_snapshot(
        r#"
    fn equals(a:int, b:int):bool                { a == b }
    fn not_equals(a:int, b:int):bool            { a != b }
    fn less(a:int, b:int):bool                  { a < b }
    fn less_equal(a:int, b:int):bool            { a <= b }
    fn greater(a:int, b:int):bool               { a > b }
    fn greater_equal(a:int, b:int):bool         { a >= b }
    fn equalsf(a:float, b:float):bool           { a == b }
    fn not_equalsf(a:float, b:float):bool       { a != b }
    fn lessf(a:float, b:float):bool             { a < b }
    fn less_equalf(a:float, b:float):bool       { a <= b }
    fn greaterf(a:float, b:float):bool          { a > b }
    fn greater_equalf(a:float, b:float):bool    { a >= b }
    "#,
    );
}

#[test]
fn if_statement() {
    test_snapshot(
        r#"
    fn foo(a:int):int {
        let b = if a > 3 {
            let c = if a > 4 {
                a+1
            } else {
                a+3
            }
            c
        } else {
            a-1
        }
        b
    }
    "#,
    )
}

#[test]
fn void_return() {
    test_snapshot(
        r#"
    fn bar() {
        let a = 3;
    }
    fn foo(a:int) {
        let c = bar()
    }
    "#,
    )
}

#[test]
fn fibonacci() {
    test_snapshot(
        r#"
    fn fibonacci(n:int):int {
        if n <= 1 {
            n
        } else {
            fibonacci(n-1) + fibonacci(n-2)
        }
    }
    "#,
    )
}

#[test]
fn fibonacci_loop() {
    test_snapshot(
        r#"
    fn fibonacci(n:int):int {
        let a = 0;
        let b = 1;
        let i = 1;
        loop {
            if i > n {
                return a
            }
            let sum = a + b;
            a = b;
            b = sum;
            i += 1;
        }
    }
    "#,
    )
}

#[test]
fn shadowing() {
    test_snapshot(
        r#"
    fn foo(a:int):int {
        let a = a+1;
        {
            let a = a+2;
        }
        a+3
    }

    fn bar(a:int):int {
        let a = a+1;
        let a = {
            let a = a+2;
            a
        }
        a+3
    }
    "#,
    );
}

#[test]
fn return_expr() {
    test_snapshot(
        r#"
    fn main():int {
        return 5;
        let a = 3; // Nothing regarding this statement should be in the IR
    }
    "#,
    );
}

#[test]
fn conditional_return_expr() {
    test_snapshot(
        r#"
    fn main(a:int):int {
        if a > 4 {
            return a;
        }
        a - 1
    }
    "#,
    );
}

#[test]
fn never_conditional_return_expr() {
    test_snapshot(
        r#"
    fn main(a:int):int {
        if a > 4 {
            return a;
        } else {
            return a - 1;
        }
    }
    "#,
    );
}

#[test]
fn true_is_true() {
    test_snapshot(
        r#"
    fn test_true():bool {
        true
    }

    fn test_false():bool {
        false
    }"#,
    );
}

#[test]
fn loop_expr() {
    test_snapshot(
        r#"
    fn foo() {
        loop {}
    }
    "#,
    )
}

#[test]
fn loop_break_expr() {
    test_snapshot(
        r#"
    fn foo(n:int):int {
        loop {
            if n > 5 {
                break n;
            }
            if n > 10 {
                break 10;
            }
            n += 1;
        }
    }
    "#,
    )
}

#[test]
fn while_expr() {
    test_snapshot(
        r#"
    fn foo(n:int) {
        while n<3 {
            n += 1;
        };

        // This will be completely optimized out
        while n<4 {
            break;
        };
       }
    "#,
    )
}

#[test]
fn struct_test() {
    test_snapshot_unoptimized(
        r#"
    struct(value) Bar(float, int, bool, Foo);
    struct(value) Foo { a: int };
    struct(value) Baz;
    fn foo() {
        let a: Foo = Foo { a: 5 };
        let b: Bar = Bar(1.23, a.a, true, a);
        let c: Baz = Baz;
    }
    "#,
    )
}

#[test]
fn field_expr() {
    test_snapshot(
        r#"
    struct(value) Bar(float, Foo);
    struct(value) Foo { a: int };

    fn bar_0(bar: Bar): float {
        bar.0
    }

    fn bar_1(bar: Bar): Foo {
        bar.1
    }

    fn bar_1_a(bar: Bar): int {
        bar.1.a
    }

    fn foo_a(foo: Foo): int {
        foo.a
    }

    fn bar_1_foo_a(bar: Bar): int {
        foo_a(bar_1(bar))
    }

    fn main(): int {
        let a: Foo = Foo { a: 5 };
        let b: Bar = Bar(1.23, a);
        let aa_lhs = a.a + 2;
        let aa_rhs = 2 + a.a;
        aa_lhs + aa_rhs
    }
    "#,
    )
}

#[test]
fn field_crash() {
    test_snapshot_unoptimized(
        r#"
    struct(gc) Foo { a: int };

    fn main(c:int):int {
        let b = Foo { a: c + 5 }
        b.a
    }
    "#,
    )
}

#[test]
fn gc_struct() {
    test_snapshot_unoptimized(
        r#"
    struct(gc) Foo { a: int, b: int };

    fn foo() {
        let a = Foo { a: 3, b: 4 };
        a.b += 3;
        let b = a;
    }
    "#,
    )
}

#[test]
fn primitive_types() {
    test_snapshot(
        r#"
   fn add(a: u8, b: u8): u8 { a+b }
   fn less(a: u8, b: u8): bool { a<b }
    "#,
    )
}

fn test_snapshot(text: &str) {
    test_snapshot_with_optimization(text, OptimizationLevel::Default);
}

fn test_snapshot_unoptimized(text: &str) {
    test_snapshot_with_optimization(text, OptimizationLevel::None);
}

fn test_snapshot_with_optimization(text: &str, opt: OptimizationLevel) {
    let text = text.trim().replace("\n    ", "\n");

    let (mut db, file_id) = MockDatabase::with_single_file(&text);
    db.set_optimization_lvl(opt);

    let line_index: Arc<LineIndex> = db.line_index(file_id);
    let messages = RefCell::new(Vec::new());
    let mut sink = DiagnosticSink::new(|diag| {
        let line_col = line_index.line_col(diag.highlight_range().start());
        messages.borrow_mut().push(format!(
            "error {}:{}: {}",
            line_col.line + 1,
            line_col.col + 1,
            diag.message()
        ));
    });
    Module::from(file_id).diagnostics(&db, &mut sink);
    drop(sink);
    let messages = messages.into_inner();

    let name = if !messages.is_empty() {
        messages.join("\n")
    } else {
        format!(
            "{}",
            db.module_ir(file_id)
                .llvm_module
                .print_to_string()
                .to_string()
        )
    };
    insta::assert_snapshot!(insta::_macro_support::AutoName, name, &text);
}
