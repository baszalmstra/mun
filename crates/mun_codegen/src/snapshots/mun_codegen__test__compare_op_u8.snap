---
source: crates/mun_codegen/src/test.rs
assertion_line: 1045
expression: "\n    pub fn equals(a: u8, b: u8) -> bool { a == b }\n    pub fn not_equal(a: u8, b: u8) -> bool { a != b}\n    pub fn less(a: u8, b: u8) -> bool { a < b }\n    pub fn less_equal(a: u8, b: u8) -> bool { a <= b }\n    pub fn greater(a: u8, b: u8) -> bool { a > b }\n    pub fn greater_equal(a: u8, b: u8) -> bool { a >= b }\n                        "

---
; == FILE IR (mod) =====================================
; ModuleID = 'mod'
source_filename = "mod"

%"mun_codegen::ir::types::TypeInfo" = type <{ [0 x i64], [16 x i8], [0 x i64], i8*, [0 x i64], i32, [0 x i64], i8, [3 x i8], %"mun_codegen::ir::types::TypeInfoData", [0 x i64] }>
%"mun_codegen::ir::types::TypeInfoData" = type <{ [0 x i8], i8, [39 x i8] }>

@global_type_table = external global [2 x %"mun_codegen::ir::types::TypeInfo"*]

define i1 @equals(i8 %0, i8 %1) {
body:
  %eq = icmp eq i8 %0, %1
  ret i1 %eq
}

define i1 @not_equal(i8 %0, i8 %1) {
body:
  %neq = icmp ne i8 %0, %1
  ret i1 %neq
}

define i1 @less(i8 %0, i8 %1) {
body:
  %less = icmp ult i8 %0, %1
  ret i1 %less
}

define i1 @less_equal(i8 %0, i8 %1) {
body:
  %lesseq = icmp ule i8 %0, %1
  ret i1 %lesseq
}

define i1 @greater(i8 %0, i8 %1) {
body:
  %greater = icmp ugt i8 %0, %1
  ret i1 %greater
}

define i1 @greater_equal(i8 %0, i8 %1) {
body:
  %greatereq = icmp uge i8 %0, %1
  ret i1 %greatereq
}

; == GROUP IR (mod) ====================================
; ModuleID = 'group_name'
source_filename = "group_name"

%"mun_codegen::ir::types::TypeInfo" = type <{ [0 x i64], [16 x i8], [0 x i64], i8*, [0 x i64], i32, [0 x i64], i8, [3 x i8], %"mun_codegen::ir::types::TypeInfoData", [0 x i64] }>
%"mun_codegen::ir::types::TypeInfoData" = type <{ [0 x i8], i8, [39 x i8] }>

@"type_info::<core::bool>::name" = private unnamed_addr constant [11 x i8] c"core::bool\00"
@"type_info::<core::bool>" = private unnamed_addr constant <{ [16 x i8], i8*, [48 x i8] }> <{ [16 x i8] c"x\82\81m t7\03\CB\F8k\81-;\C9\84", i8* getelementptr inbounds ([11 x i8], [11 x i8]* @"type_info::<core::bool>::name", i32 0, i32 0), [48 x i8] c"\01\00\00\00\01\00\00\00\00\00\00\00\00\00\00\00\00\00\00\00\00\00\00\00\00\00\00\00\00\00\00\00\00\00\00\00\00\00\00\00\00\00\00\00\00\00\00\00" }>
@"type_info::<core::u8>::name" = private unnamed_addr constant [9 x i8] c"core::u8\00"
@"type_info::<core::u8>" = private unnamed_addr constant <{ [16 x i8], i8*, [48 x i8] }> <{ [16 x i8] c"\A0y\A7S\B6(n\F7f&H\E1\F9\AD\04>", i8* getelementptr inbounds ([9 x i8], [9 x i8]* @"type_info::<core::u8>::name", i32 0, i32 0), [48 x i8] c"\08\00\00\00\01\00\00\00\00\00\00\00\00\00\00\00\00\00\00\00\00\00\00\00\00\00\00\00\00\00\00\00\00\00\00\00\00\00\00\00\00\00\00\00\00\00\00\00" }>
@global_type_table = constant [2 x %"mun_codegen::ir::types::TypeInfo"*] [%"mun_codegen::ir::types::TypeInfo"* @"type_info::<core::bool>", %"mun_codegen::ir::types::TypeInfo"* @"type_info::<core::u8>"]

